### PR TITLE
[6CORE-024] Inventory Management

### PR DESCRIPTION
**Pull request by VictoriaKh-T**

### CODE REVIEW

File: 3-platform/warehouse/spec/src/main/java/com/_6core/platform/warehousespec/rest/v1/InventorySupplyApi.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+import reactor.core.publisher.Mono;
+
+/**
+ * Interface for supplying new products to the inventory in the Warehouse Service.
+ *
+ * This interface defines a single method for supplying new products to the inventory.
+ */
+
+@Tag(name = "Inventory supplying",
+        description = "REST specification for supplying ")
```
Comment (kateryna-mykh) [inline 3-platform/warehouse/spec/src/main/java/com/_6core/platform/warehousespec/rest/v1/InventorySupplyApi.java]: Space in the end of description

File: 3-platform/warehouse/spec/src/main/java/com/_6core/platform/warehousespec/rest/v1/InventorySupplyApi.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+     *             - 404 (Not Found) - if the inventory item with the provided ID is not found.
+     *             - 500 (Internal Server Error) - if an unexpected error occurs during the supply process.
+     */
+    @Operation(summary = "Supply products to inventory", description = "Supplying new products to inventory")
+    @ApiResponses(value = {
+            @ApiResponse(responseCode = "200", description = "OK", content = {
+                    @Content(mediaType = "application/json",
+                            schema = @Schema(implementation = InventorySupplyResponse.class))}),
+            @ApiResponse(responseCode = "404", description = "inventory not found"),
+            @ApiResponse(responseCode = "500", description = "Internal server error")
```
Comment (kateryna-mykh) [inline 3-platform/warehouse/spec/src/main/java/com/_6core/platform/warehousespec/rest/v1/InventorySupplyApi.java]: description difference, (I meant one start with lowercase and other with uppercase)

File: 3-platform/warehouse/spec/src/main/java/com/_6core/platform/warehousespec/rest/v1/InventoryTransferAPI.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+import org.springframework.web.bind.annotation.RequestBody;
+import reactor.core.publisher.Mono;
+
+/**
+ * Interface for allocation inventory in the Warehouse Service.
+ */
+
+@Tag(name = "Inventory allocate",
+        description = "REST specification for allocation ")
+public interface InventoryTransferAPI {
```
Comment (kateryna-mykh) [inline 3-platform/warehouse/spec/src/main/java/com/_6core/platform/warehousespec/rest/v1/InventoryTransferAPI.java]: One file is "..Api" other is "..API"

File: 3-platform/warehouse/spec/src/main/java/com/_6core/platform/warehousespec/rest/v1/InventoryTransferAPI.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.RequestBody;
+import reactor.core.publisher.Mono;
+
+/**
+ * Interface for allocation inventory in the Warehouse Service.
+ */
+
+@Tag(name = "Inventory allocate",
+        description = "REST specification for allocation ")
```
Comment (kateryna-mykh) [inline 3-platform/warehouse/spec/src/main/java/com/_6core/platform/warehousespec/rest/v1/InventoryTransferAPI.java]: whitespace in the end of description

File: 3-platform/warehouse/spec/src/main/java/com/_6core/platform/warehousespec/rest/v1/InventoryTransferAPI.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+            @ApiResponse(responseCode = "200", description = "OK", content = {
+                    @Content(mediaType = "application/json",
+                            schema = @Schema(implementation = InventoryTransferResponse.class))}),
+            @ApiResponse(responseCode = "404", description = "inventory not found"),
+            @ApiResponse(responseCode = "500", description = "Internal server error")
+    })
+    @PatchMapping("invetory/{inventoryId}/transfer")
+    default Mono<ResponseEntity<InventoryTransferResponse>> transferInventory (
+            @RequestBody InventoryTransferRequest request,
+            @PathVariable String inventoryId) {
```
Comment (kateryna-mykh) [inline 3-platform/warehouse/spec/src/main/java/com/_6core/platform/warehousespec/rest/v1/InventoryTransferAPI.java]: in method above you add @Parameter(description  = "inventory ID") for @PathVariable, but not there, why?

File: 3-platform/warehouse/spec/src/main/java/com/_6core/platform/warehousespec/rest/v1/dto/inventory/InventorySupplyRequest.java [hunk 6]
```diff
@@ -0,0 +1,6 @@
+package com._6core.platform.warehousespec.rest.v1.dto.inventory;
+
+public record InventorySupplyRequest(String inventoryId,
+                                     String productId,
+                                     Integer quality) {
```
Comment (kateryna-mykh) [inline 3-platform/warehouse/spec/src/main/java/com/_6core/platform/warehousespec/rest/v1/dto/inventory/InventorySupplyRequest.java]: in that record and in the others: is that should be quality or rather quantity?

**Review by kateryna-mykh — CHANGES_REQUESTED** [review]

File: 3-platform/warehouse/spec/src/main/java/com/_6core/platform/warehousespec/rest/v1/InventorySupplyApi.java [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    })
+    @PatchMapping("inventory/{inventoryId}/supply")
```
Comment (sokimaaa) [inline 3-platform/warehouse/spec/src/main/java/com/_6core/platform/warehousespec/rest/v1/InventorySupplyApi.java]: REST URI naming violation. pls check https://restfulapi.net/resource-naming/

File: 3-platform/warehouse/spec/src/main/java/com/_6core/platform/warehousespec/rest/v1/InventoryTransferApі.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    @Operation(summary = "Transfer inventory", description = "Moving inventory in Warehouse")
+    @ApiResponses(value = {
+            @ApiResponse(responseCode = "200", description = "OK", content = {
+                    @Content(mediaType = "application/json",
+                            schema = @Schema(implementation = InventoryTransferResponse.class))}),
+            @ApiResponse(responseCode = "404", description = "inventory not found"),
+            @ApiResponse(responseCode = "500", description = "Internal server error")
+    })
+    @PatchMapping("invetory/{inventoryId}/transfer")
+    default Mono<ResponseEntity<InventoryTransferResponse>> transferInventory (
```
Comment (sokimaaa) [inline 3-platform/warehouse/spec/src/main/java/com/_6core/platform/warehousespec/rest/v1/InventoryTransferApі.java]: Not Transfering! `Allocation` is about creation of a new inventory spot for on merchant' request. 
```suggestion
    default Mono<ResponseEntity<InventoryTransferResponse>> transferInventory (
```

**Review by sokimaaa — CHANGES_REQUESTED** [review]

File: 3-platform/warehouse/spec/src/main/java/com/_6core/platform/warehousespec/rest/v1/InventoryTransferApі.java (hunk 8 above)
Comment (sokimaaa) [inline 3-platform/warehouse/spec/src/main/java/com/_6core/platform/warehousespec/rest/v1/InventoryTransferApі.java]: BTW: I think that `allocation` and `supply` could be merged into one interface. It is all about `management` of inventory. No need to create a new class per method.

**Review by sokimaaa — CHANGES_REQUESTED** [review]

**Review by kateryna-mykh — APPROVED** [review]

File: 3-platform/warehouse/spec/src/main/java/com/_6core/platform/warehousespec/rest/v1/InventoryManagementApi.java [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+    @ApiResponses(value = {
+            @ApiResponse(responseCode = "200", description = "OK", content = {
+                    @Content(mediaType = "application/json",
+                            schema = @Schema(implementation = InventoryAllocateResponse.class))}),
+            @ApiResponse(responseCode = "400", description = "Field validation error"),
+            @ApiResponse(responseCode = "401", description = "Authentication error"),
+            @ApiResponse(responseCode = "404", description = "inventory not found"),
+            @ApiResponse(responseCode = "500", description = "Internal server error")
+    })
+    @PostMapping("/inventory/allocate")
```
Comment (sokimaaa) [inline 3-platform/warehouse/spec/src/main/java/com/_6core/platform/warehousespec/rest/v1/InventoryManagementApi.java]: Fix URI regarding REST URI naming Conventions for both allocation and supply.
https://restfulapi.net/resource-naming/

https://medium.com/@nadinCodeHat/rest-api-naming-conventions-and-best-practices-1c4e781eb6a5
```suggestion
    @PostMapping("/inventory/allocate")
```

**Review by sokimaaa — CHANGES_REQUESTED** [review]

File: 3-platform/warehouse/spec/src/main/java/com/_6core/platform/warehousespec/rest/v1/InventoryManagementApi.java [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+    @ApiResponses(value = {
+            @ApiResponse(responseCode = "200", description = "OK", content = {
+                    @Content(mediaType = "application/json",
+                            schema = @Schema(implementation = InventoryAllocateResponse.class))}),
+            @ApiResponse(responseCode = "400", description = "Field validation error"),
+            @ApiResponse(responseCode = "401", description = "Authentication error"),
+            @ApiResponse(responseCode = "404", description = "inventory not found"),
+            @ApiResponse(responseCode = "500", description = "Internal server error")
+    })
+    @PostMapping("/inventories/{inventoryId}/allocations")
```
Comment (sokimaaa) [inline 3-platform/warehouse/spec/src/main/java/com/_6core/platform/warehousespec/rest/v1/InventoryManagementApi.java]: REST typically provides the resources, e.g. inventories. This is why we name it in the plural.
Is `allocations` the resource? I don't think so. 

`Allocate` is verb and URI convention recommends getting rid of `verbs`. Why? We have methods like POST, PATCH, PUT ,... that say to us what endpoint is going to do. 
So adding the verb is redundant in this case. 

The same for other endpoints.
```suggestion
    @PostMapping("/inventories/{inventoryId}")
```

Comment (sokimaaa) [inline 3-platform/warehouse/spec/src/main/java/com/_6core/platform/warehousespec/rest/v1/InventoryManagementApi.java]: you forgot about `warehouses/{warehouseId}`. Merchant is going to allocate space in the specific warehouse, isn't it?

File: 3-platform/warehouse/spec/src/main/java/com/_6core/platform/warehousespec/rest/v1/InventoryManagementApi.java [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+    @ApiResponses(value = {
+            @ApiResponse(responseCode = "200", description = "OK", content = {
+                    @Content(mediaType = "application/json",
+                            schema = @Schema(implementation = InventorySupplyResponse.class))}),
+            @ApiResponse(responseCode = "400", description = "Field validation error"),
+            @ApiResponse(responseCode = "401", description = "Authentication error"),
+            @ApiResponse(responseCode = "404", description = "Inventory not found"),
+            @ApiResponse(responseCode = "500", description = "Internal server error")
+    })
+    @PatchMapping("/inventories/{inventoryId}/products")
```
Comment (sokimaaa) [inline 3-platform/warehouse/spec/src/main/java/com/_6core/platform/warehousespec/rest/v1/InventoryManagementApi.java]: Are we changing inventory or products? I think inventory.
You can remember, that DELETE, PUT, and PATCH methods apply only to a singular resource.
That means the last element in URI is going to be `somes/{someId}`

```suggestion
    @PatchMapping("/inventories/{inventoryId}")
```

**Review by sokimaaa — CHANGES_REQUESTED** [review]

File: 3-platform/warehouse/spec/src/main/java/com/_6core/platform/warehousespec/rest/v1/InventoryManagementApi.java (hunk 10 above)
Comment (sokimaaa) [reply 3-platform/warehouse/spec/src/main/java/com/_6core/platform/warehousespec/rest/v1/InventoryManagementApi.java]: the same for supply. We need to have information about which warehouse was supplied

**Review by sokimaaa — APPROVED** [review]

**Pull request merged**